File: 2024.1/lab1/python/concurrent/sum.py

import sys
import threading

semaphore = threading.Semaphore(1)
total_sum = 0

def do_sum(path):
    global total_sum
    _sum = 0
    with open(path, 'rb',buffering=0) as f:
        byte = f.read(1)
        while byte:
            _sum += int.from_bytes(byte, byteorder='big', signed=False)
            byte = f.read(1)

    semaphore.acquire()
    try:
        total_sum += _sum
    finally:
        semaphore.release()
    
    print(path + " : " + str(_sum))
    


if __name__ == "__main__":
    paths = sys.argv[1:]
    threads = []
    for path in paths:
        thread = threading.Thread(target=do_sum, args=(path,))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    print(f"Soma total de todos os arquivos: {total_sum}")

File: 2024.1/lab1/java/concurrent/Sum.java
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.Semaphore;

public class Sum {

    private static long totalSum = 0;
    private static final Semaphore semaphore = new Semaphore(1);

    public static int sum(FileInputStream fis) throws IOException {
        int sum = 0;
        int byteRead = fis.read();
        while (byteRead != -1) {
        	sum += byteRead;
            byteRead = fis.read();
        }
        return sum;
    }

    public static long sum(String path) throws IOException {

        Path filePath = Paths.get(path);
        if (Files.isRegularFile(filePath)) {
       	    FileInputStream fis = new FileInputStream(filePath.toString());
            return sum(fis);
        } else {
            throw new RuntimeException("Non-regular file: " + path);
        }
    }

    public static void main(String[] args) throws Exception {

        if (args.length < 1) {
            System.err.println("Usage: java Sum filepath1 filepath2 filepathN");
            System.exit(1);
        }

        Thread[] threads = new Thread[args.length];

        for(int i=0; i < args.length; i++){
            String path = args[i];

            Runnable task = new Runnable(){
                @Override
                public void run(){
                    try {
                        long fileSum = sum(path); // Soma para o arquivo atual

                        // Região crítica protegida por semáforo
                        semaphore.acquire(); // Bloqueia o acesso
                        try {
                            totalSum += fileSum; // Modifica a soma total
                        } finally {
                            semaphore.release(); // Libera o acesso
                        }

                        System.out.println(path + " : " + fileSum);
                    } catch (IOException e) {
                        System.err.println("Erro ao processar " + path + ": " + e.getMessage());
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        System.err.println("Thread interrompida: " + e.getMessage());
                    }
                }
            };
        }

        for(Thread thread: threads){
            thread.join();
        }
        System.out.println("Soma total de todos os arquivos: " + totalSum);

    }



}
